File: backend/app/schemas/application.py
from datetime import datetime
from typing import Any
from uuid import UUID

from pydantic import Field, field_validator

from app.models.application import AppStatus
from app.schemas.common import BaseSchema


class ApplicationBase(BaseSchema):
    """Base application schema."""

    name: str = Field(..., min_length=1, max_length=255)
    description: str | None = None
    logo_url: str | None = None
    icon: str | None = None
    base_url: str = Field(..., description="Base URL for the application API")
    callback_url: str | None = None
    launch_url: str | None = None
    features_manifest_url: str | None = Field(
        None,
        description="Custom URL for features manifest. If empty, uses {base_url}/api/v1/app-features/manifest",
    )
    healthcheck_url: str | None = None
    auth_mode: str = Field(default="platform_jwt")
    metadata_: dict = Field(default_factory=dict, alias="metadata")

    @field_validator("metadata_", mode="before")
    @classmethod
    def ensure_dict(cls, v: Any) -> dict:
        if v is None:
            return {}
        if isinstance(v, dict):
            return v
        return dict(v) if hasattr(v, "__iter__") else {}


class ApplicationCreate(ApplicationBase):
    """Schema for registering a new application."""

    id: str = Field(
        ...,
        min_length=1,
        max_length=100,
        pattern=r"^[a-z0-9_]+$",
        description="Unique application identifier (e.g., orchestrator_ai)",
    )
    app_catalog_id: str = Field(..., description="App catalog entry ID")
    tenant_id: UUID | None = Field(None, description="Tenant ID (defaults to token tenant)")
    # Override: name is optional on create — falls back to catalog name
    name: str | None = Field(None, max_length=255)


class ApplicationUpdate(BaseSchema):
    """Schema for updating an application."""

    name: str | None = Field(None, min_length=1, max_length=255)
    description: str | None = None
    logo_url: str | None = None
    icon: str | None = None
    base_url: str | None = None
    callback_url: str | None = None
    launch_url: str | None = None
    features_manifest_url: str | None = None
    healthcheck_url: str | None = None
    status: AppStatus | None = None
    auth_mode: str | None = None
    metadata_: dict | None = Field(None, alias="metadata")


class ApplicationRead(ApplicationBase):
    """Schema for reading application data."""

    id: str
    tenant_id: UUID
    app_catalog_id: str | None = None
    status: AppStatus
    current_version: str | None = None
    created_at: datetime
    updated_at: datetime

    # Computed fields
    permissions_count: int | None = None
    features_count: int | None = None
    last_sync_at: datetime | None = None
    sync_status: str | None = None


class ApplicationSummary(BaseSchema):
    """Minimal application info."""

    id: str
    name: str
    status: AppStatus
    current_version: str | None = None


# Tenant Application schemas
class TenantApplicationBase(BaseSchema):
    """Base schema for tenant-application relationship."""

    config: dict = Field(default_factory=dict)

    @field_validator("config", mode="before")
    @classmethod
    def ensure_config_dict(cls, v: Any) -> dict:
        if v is None:
            return {}
        if isinstance(v, dict):
            return v
        return dict(v) if hasattr(v, "__iter__") else {}


class TenantApplicationCreate(TenantApplicationBase):
    """Schema for enabling an application for a tenant."""

    application_id: str


class TenantApplicationUpdate(BaseSchema):
    """Schema for updating tenant-application settings."""

    status: AppStatus | None = None
    config: dict | None = None


class TenantApplicationRead(TenantApplicationBase):
    """Schema for reading tenant-application data."""

    id: UUID
    tenant_id: UUID
    application_id: str
    status: AppStatus
    enabled_at: datetime | None = None
    disabled_at: datetime | None = None
    created_at: datetime
    updated_at: datetime

    # Nested application info
    application: ApplicationSummary | None = None


# Permission sync schemas
class PermissionSyncRequest(BaseSchema):
    """Request to trigger permission sync."""

    force: bool = Field(default=False, description="Force sync even if up to date")


class PermissionSyncResponse(BaseSchema):
    """Response from permission sync."""

    status: str
    app_version: str | None = None
    summary: dict = Field(default_factory=dict)
    error_message: str | None = None


class AppLauncherItem(BaseSchema):
    """Application entry shown in TAH app launcher."""

    id: str
    name: str
    description: str | None = None
    icon: str | None = None
    logo_url: str | None = None
    base_url: str
    launch_url: str | None = None
    callback_url: str | None = None
    status: AppStatus
    category: str | None = None


class AppLauncherResponse(BaseSchema):
    """Launcher payload scoped to current user and tenant."""

    tenant_id: str
    tenant_name: str
    current_user_id: str | None = None
    current_user_name: str | None = None
    current_user_email: str | None = None
    current_user_roles: list[str] = Field(default_factory=list)
    can_access_admin: bool = False
    applications: list[AppLauncherItem] = Field(default_factory=list)

File: frontend/src/types/application.ts
export type AppStatus = "active" | "inactive" | "maintenance" | "deprecated"
export type CatalogStatus = "active" | "inactive"
export type AppCategory = "studio" | "production" | "governance" | "data" | "settings"

// App Catalog - Global dictionary of applications
export interface AppCatalog {
  id: string
  name: string
  description: string | null
  logo_url: string | null
  category: AppCategory | null
  status: CatalogStatus
  created_at: string
  updated_at: string
  tenant_count?: number
}

export interface AppCatalogCreate {
  id: string
  name: string
  description?: string
  logo_url?: string
  category?: AppCategory
  status?: CatalogStatus
}

export interface AppCatalogUpdate {
  name?: string
  description?: string
  logo_url?: string
  category?: AppCategory
  status?: CatalogStatus
}

export interface AppCatalogList {
  id: string
  name: string
  description: string | null
  logo_url: string | null
  category: AppCategory | null
  status: CatalogStatus
}

// Application - Tenant-specific configuration
export interface Application {
  id: string
  tenant_id: string | null
  app_catalog_id: string | null
  name: string
  description: string | null
  base_url: string
  features_manifest_url: string | null
  status: AppStatus
  current_version: string | null
  healthcheck_url: string | null
  icon: string | null
  callback_url: string | null
  launch_url: string | null
  logo_url: string | null
  auth_mode: string
  metadata: Record<string, unknown>
  created_at: string
  updated_at: string
  permissions_count?: number
  features_count?: number
  last_sync_at?: string | null
  sync_status?: string | null
}

export interface ApplicationCreate {
  id: string
  app_catalog_id: string  // Required - must select from catalog
  base_url: string
  features_manifest_url?: string
  healthcheck_url?: string
  icon?: string
  callback_url?: string
  launch_url?: string
  auth_mode?: string
  metadata?: Record<string, unknown>
  // Optional overrides (if not provided, uses catalog values)
  name?: string
  description?: string
  logo_url?: string
}

export interface ApplicationUpdate {
  app_catalog_id?: string  // Can change the catalog app
  name?: string
  description?: string
  base_url?: string
  features_manifest_url?: string
  healthcheck_url?: string
  icon?: string
  callback_url?: string
  launch_url?: string
  logo_url?: string
  status?: AppStatus
  auth_mode?: string
  metadata?: Record<string, unknown>
}

export interface TenantApplication {
  id: string
  tenant_id: string
  application_id: string
  status: AppStatus
  enabled_at: string | null
  disabled_at: string | null
  config: Record<string, unknown>
  created_at: string
  updated_at: string
  application?: Application
}
